### PR TITLE
[SYCLomatic] Migration of cudaGraphNode_t,cudaGraphAddEmptyNode,cudaGraphAddDependencies

### PR DESCRIPTION
**Pull request by daiyaan-ahmed6**
Signed-off-by: Daiyaan Ahmed <daiyaan.ahmed@intel.com>

### CODE REVIEW

File: clang/test/dpct/cudaGraph_test.cu [hunk 1]
```diff
@@ -33,6 +33,30 @@ int main() {
...
+  cudaGraphNode_t node6, *node7, **node8;
+
+  // CHECK: dpct::experimental::add_empty_node(&node, graph, node4, 10);
+  cudaGraphAddEmptyNode(&node, graph, node4, 10);
+
+  // CHECK: dpct::experimental::add_empty_node(node2, *graph2, NULL, 0);
+  cudaGraphAddEmptyNode(node2, *graph2, NULL, 0);
+
+  // CHECK: dpct::experimental::add_dependencies(graph, node4, node5, 10);
+  cudaGraphAddDependencies(graph, node4, node5, 10);
```
Comment (tangjj11) [inline clang/test/dpct/cudaGraph_test.cu:68]: Need test some args with `const` qualifier, for example:
```c++
const cudaGraphNode_t* node4;
const cudaGraphNode_t* node5;
cudaGraphAddDependencies(graph, node4, node5, 10);
```
If build failed, we need add `const` of param in helper functions.

**Review by tangjj11 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 2]
```diff
@@ -22,6 +22,8 @@ typedef sycl::ext::oneapi::experimental::command_graph<
     sycl::ext::oneapi::experimental::graph_state::executable>
     *command_graph_exec_ptr;
 
+typedef std::shared_ptr<sycl::ext::oneapi::experimental::node> node_ptr;
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: As raw pointer is used for graph, I suggest using raw pointer here too.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 3]
```diff
@@ -89,5 +91,47 @@ static inline bool end_recording(sycl::queue *queue_ptr,
...
+/// representing the dependencies of the new node.
+/// \param [in] numberOfDependencies The number of dependencies in
+/// the dependenciesArray.
+static void add_empty_node(dpct::experimental::node_ptr *newNode,
+                           dpct::experimental::command_graph_ptr graph,
+                           dpct::experimental::node_ptr *dependenciesArray,
+                           std::size_t numberOfDependencies) {
+  *newNode =
+      std::make_shared<sycl::ext::oneapi::experimental::node>(graph->add());
+  if (numberOfDependencies) {
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: Can we use one property `depends_on` which contains all the dependencies, or add multiple properties in one call?

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 4]
```diff
@@ -89,5 +91,47 @@ static inline bool end_recording(sycl::queue *queue_ptr,
...
+/// \param [in] dependenciesArray An array of node pointers
+/// representing the dependencies of the new node.
+/// \param [in] numberOfDependencies The number of dependencies in
+/// the dependenciesArray.
+static void add_empty_node(dpct::experimental::node_ptr *newNode,
+                           dpct::experimental::command_graph_ptr graph,
+                           dpct::experimental::node_ptr *dependenciesArray,
+                           std::size_t numberOfDependencies) {
+  *newNode =
+      std::make_shared<sycl::ext::oneapi::experimental::node>(graph->add());
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: What is the purpose for this line?

File: clang/test/dpct/cudaGraph_test.cu (hunk 1 above)
Comment (daiyaan-ahmed6) [reply clang/test/dpct/cudaGraph_test.cu:68]: Done.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp (hunk 2 above)
Comment (daiyaan-ahmed6) [reply clang/runtime/dpct-rt/include/dpct/graph.hpp]: Done.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp (hunk 3 above)
Comment (daiyaan-ahmed6) [reply clang/runtime/dpct-rt/include/dpct/graph.hpp]: Added logic to add dependencies at once.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp (hunk 4 above)
Comment (daiyaan-ahmed6) [reply clang/runtime/dpct-rt/include/dpct/graph.hpp]: This adds an empty node without any dependencies. Logic updated to be more readable in latest commit.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 5]
```diff
@@ -82,12 +84,57 @@ static inline bool begin_recording(sycl::queue *queue_ptr) {
...
+               const dpct::experimental::node_ptr *dependenciesArray,
+               std::size_t numberOfDependencies) {
+  if (numberOfDependencies == 0) {
+    *newNode = new sycl::ext::oneapi::experimental::node(graph->add());
+    return;
+  }
+  std::vector<sycl::ext::oneapi::experimental::node> dependencies;
+  for (std::size_t i = 0; i < numberOfDependencies; i++) {
+    dependencies.push_back(*dependenciesArray[i]);
+  }
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp:115]:
```suggestion
  std::vector<sycl::ext::oneapi::experimental::node> dependencies{dependenciesArray, dependenciesArray + numberOfDependencies};
```

Comment (daiyaan-ahmed6) [reply clang/runtime/dpct-rt/include/dpct/graph.hpp:115]: Done.

**Review by ziranzha — APPROVED** [review]

**Pull request merged**